Ability to switch between different Locales and Currencies in a single Store

damianlegawiec opened the pull request:
https://user-images.githubusercontent.com/55154/107916015-d1f7f380-6f65-11eb-8345-c70e8c4cee10.mov

damianlegawiec committed:
Added Store#supported_locales

This is the backbone for multi currency inside a single Store

damianlegawiec committed:
Add Supported Locales in Admin Panel UI

damianlegawiec committed:
Modify Store dropdown

damianlegawiec committed:
Additional tests

spencerwp commented:
This is great - I'm only seeing one problem where :supported_locales is not a valid parameter of the store_attributes.

spree/core/lib/spree/permitted_attributes.rb - 
@@store_attributes += [:supported_locales]

damianlegawiec committed:
Ensure supported locales / currencies are set for Store

damianlegawiec committed:
Expose supported_locales in API v1 and v2

flobby commented:
Not sure if related but on admin backend when I try to access store translations I receive this error. Might be related to globalize though.

NoMethodError - undefined method `translated_attribute_names' for #<Class:0x00007fefa2c5c580>:


I tracked it down to admin/translations_controller.rb:10 // something in the index action which is very generic.

(using master on commit 166b5434860636f7fb36beb3eaa6a2763b854077)

damianlegawiec commented:
> Not sure if related but on admin backend when I try to access store translations I receive this error. Might be related to globalize though.
> 
> NoMethodError - undefined method `translated_attribute_names' for #Class:0x00007fefa2c5c580:
> 
> I tracked it down to admin/translations_controller.rb:10 // something in the index action which is very generic.
> 
> (using master on commit [166b543](https://github.com/spree/spree/commit/166b5434860636f7fb36beb3eaa6a2763b854077))

Alongside this PR I'm preparing PRs for Spree Globalize and Spree I18n ( https://github.com/spree-contrib/spree_i18n/pull/876 ) to ease the migration path, hopefully, I can finish all of it today/Monday.

channainfo commented:
@damianlegawiec awesome! thank you

damianlegawiec committed:
Bring back Locale controller from Spree I18n

* adds ability to switch locale within scope of current store
* endpoint to render dropdown with a list supported currencies
* additional tests

damianlegawiec committed:
Always use upcased version of currency (iso format)

damianlegawiec committed:
Always sort supported locale/currency alphabetically

damianlegawiec committed:
Move Admin::CurrencyHelper to CurrencyHelper

To be able to use it in Storefront as well

damianlegawiec committed:
Always use current_currency

`params[:currency]` can pass not supported currency in current store.

damianlegawiec committed:
Tests for Store Helper

damianlegawiec committed:
Added StoreHelper#should_render_store_chooser?

damianlegawiec committed:
Make Store Selected hidden by default

Most people will use single store with multiple currencies / locales

damianlegawiec committed:
Added missing LocaleHelper specs

damianlegawiec committed:
Added LocaleHelper#should_render_locale_dropdown?

damianlegawiec committed:
Added CurencyHelper#should_render_currency_dropdown?

damianlegawiec committed:
Added NavigationHelper#should_render_internationalization_dropdown?

damianlegawiec committed:
Add UI for switching Locale / Currency in a single Store

This is compatible with Spree I18n / Spree Multi Currency behaviour and is best suited for single store installations that want one store on a single domain with multiple currencies / locales

damianlegawiec committed:
Use guard clause for scenarios where current_store is not defined

Fallback to `Spree::Config[:currency]` as a last resort

damianlegawiec committed:
Fixed Store Configuration spec

damianlegawiec committed:
Fix API v2 spec for not supported currencies

damianlegawiec committed:
Update 4.2 release notes

damianlegawiec commented:
> This is great - I'm only seeing one problem where :supported_locales is not a valid parameter of the store_attributes.
> 
> spree/core/lib/spree/permitted_attributes.rb -
> @@store_attributes += [:supported_locales]

Done and done!

Vegann commented on frontend/app/helpers/spree/store_helper.rb:
```diff
@@ -0,0 +1,35 @@
+    def store_country_iso(store)
+      store ||= current_store
+      return unless store
+      return unless store.default_country
```
Maybe would be better to rewrite this 2 line like this
```suggestion
      return unless store&.default_country
```

damianlegawiec committed:
Added FrontendHelper#country_flag_icon

damianlegawiec committed:
Added StoreHelper#store_link

damianlegawiec committed:
Added mobile Currency / Locale switcher UI

damianlegawiec committed:
Added Spree.showProgressBar
